### DISCUSSION TITLE
TypeError while using socket.bind()

### DISCUSSION THREAD

**Discussion by ziGuy** [micropython/micropython#12569] [ESP32] [answered]
Hello,
I've cerate an [issue](https://github.com/micropython/micropython-lib/issues/744) as well. I'm not sure which is more relevant - please feel free to remove one.

I'm trying to code a webserver using usocket lib but got the following issue.
Can you please guide me to fix this issue? is it a bug?

### My environment:
micropython -v
MicroPython v1.19.1-859-g41ed01f13 on 2023-02-13; linux [GCC 6.3.0] version

This happens as well on my esp32 device, which i believe running micropython 1.20

### My code:
```
import usocket as socket

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.bind(('0.0.0.0', 80))
```

### Traceback:
```
Traceback (most recent call last):
  File "boot.py", line 4, in <module>
TypeError: object with buffer protocol required
```

**Edit:**
* I've tried with latest micropython and issue still occurs:
MicroPython v1.20.0-544-g977dc9a36 on 2023-10-02; linux [GCC 6.3.0] version
* Replacing bind function with the following didn't help:
s.bind((**b**'0.0.0.0', 80))

**Comment by shariltumin** [comment, marked as answer]
Please try:

```
>>> import socket
>>> addr = socket.getaddrinfo('0.0.0.0', 80)[0][-1]
>>> s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
>>> s.bind(addr)
```

**Comment by ziGuy** [comment]
Now bind gives:
`OSError: [Errno 13] EACCES`

When i add:
`print(f"addr: {addr}")`
`print(f"type of addr: {type(addr)}")`

I get:
`addr: bytearray(b'\x02\x00\x00P\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00')`
`type of addr: <class 'bytearray'>`

**Comment by shariltumin** [comment]
On Linux, ports below 1000 are reserved for the system and can only be accessed by root. Please try port 8080.

**Comment by ziGuy** [comment]
This worked.
This is a bit weird for me as a web server should not run as root and it should run in port 80.

Looks like there is something is wrong with my infrastructure as now i got the following, when trying to use micropyserver:
...micropyserver/micropyserver.py", line 91, in send
AttributeError: 'socket' object has no attribute 'sendall'

But this is off-topic so I'll close this thread and try to figure out what's going on.
Thank you very much for your help.
